src/number_theory/modular.lean

```diff
@@ -318,19 +338,63 @@ end
 /-- The matrix `T = [[1,1],[0,1]]` as an element of `SL(2,ℤ)` -/
 def T : SL(2,ℤ) := ⟨![![1, 1], ![0, 1]], by norm_num [matrix.det_fin_two]⟩
 
-/-- The matrix `T' (= T⁻¹) = [[1,-1],[0,1]]` as an element of `SL(2,ℤ)` -/
-def T' : SL(2,ℤ) := ⟨![![1, -1], ![0, 1]], by norm_num [matrix.det_fin_two]⟩
-
 /-- The matrix `S = [[0,-1],[1,0]]` as an element of `SL(2,ℤ)` -/
 def S : SL(2,ℤ) := ⟨![![0, -1], ![1, 0]], by norm_num [matrix.det_fin_two]⟩
 
-/-- The standard (closed) fundamental domain of the action of `SL(2,ℤ)` on `ℍ` -/
-def fundamental_domain : set ℍ :=
-{z | 1 ≤ (complex.norm_sq z) ∧ |z.re| ≤ (1 : ℝ) / 2}
+lemma coe_S : ↑ₘS = ![![0, -1], ![1, 0]] := rfl
 
-localized "notation `𝒟` := modular_group.fundamental_domain" in modular
+lemma coe_T : ↑ₘT = ![![1, 1], ![0, 1]] := rfl
 
-/-- If `|z|<1`, then applying `S` strictly decreases `im` -/
+lemma coe_T_inv : ↑ₘ(T⁻¹) = ![![1, -1], ![0, 1]] := by simp [coe_inv, coe_T, adjugate_fin_two]
+
+lemma coe_T_zpow (n : ℤ) : ↑ₘ(T ^ n) = ![![1, n], ![0,1]] :=
+begin
+  induction n using int.induction_on with n h n h,
+  { ext i j, fin_cases i; fin_cases j;
+    simp, },
+  { rw [zpow_add, zpow_one, coe_mul, h, coe_T],
+    ext i j, fin_cases i; fin_cases j;
+    simp [matrix.mul_apply, fin.sum_univ_succ, add_comm (1 : ℤ)], },
+  { rw [zpow_sub, zpow_one, coe_mul, h, coe_T_inv],
+    ext i j, fin_cases i; fin_cases j;
+    simp [matrix.mul_apply, fin.sum_univ_succ, neg_add_eq_sub (1 : ℤ)], },
+end
+
+@[simp] lemma coe_T_zpow_smul_eq {z : ℍ} {n : ℤ} : (↑((T^n) • z) : ℂ) = z + n :=
+by simp [coe_T_zpow]
+
+/-- If instead we had `g` and `T` of type `PSL(2, ℤ)`, then we could simply state `g = T^n`. -/
+lemma exists_eq_T_zpow_of_c_eq_zero {g : SL(2, ℤ)} (hc : ↑ₘg 1 0 = 0) :
+  ∃ (n : ℤ), ∀ (z : ℍ), g • z = T^n • z :=
+begin
+  have had := g.det_coe,
+  replace had : ↑ₘg 0 0 * ↑ₘg 1 1 = 1, { rw [det_fin_two, hc] at had, linarith, },
+  rcases int.eq_one_or_neg_one_of_mul_eq_one' had with ⟨ha, hd⟩ | ⟨ha, hd⟩,
+  { use ↑ₘg 0 1,
+    suffices : g = T^(↑ₘg 0 1), { intros z, conv_lhs { rw this, }, },
+    ext i j, fin_cases i; fin_cases j;
+    simp [ha, hc, hd, coe_T_zpow], },
+  { use -↑ₘg 0 1,
+    suffices : g = -T^(-↑ₘg 0 1), { intros z, conv_lhs { rw [this, neg_smul], }, },
+    ext i j, fin_cases i; fin_cases j;
+    simp [ha, hc, hd, coe_T_zpow], },
+end
+
+/- If `c = 1`, then `g` factorises into a product terms involving only `T` and `S`. -/
+lemma g_eq_of_c_eq_one {g : SL(2,ℤ)} (hc : ↑ₘg 1 0 = 1) :
+  g = T^(↑ₘg 0 0) * S * T^(↑ₘg 1 1) :=
+begin
+  have hg := g.det_coe.symm,
+  replace hg : ↑ₘg 0 1 = ↑ₘg 0 0 * ↑ₘg 1 1 - 1, { rw [det_fin_two, hc] at hg, linarith, },
+  ext i j, fin_cases i; fin_cases j;
+  simp [coe_S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ, hg, hc],
+end
+
+/-- If `1 < |z|`, then `|S•z| < 1` -/
+lemma norm_sq_S_smul_lt_one {z : ℍ} (h: 1 < norm_sq z) : norm_sq ↑(S • z) < 1 :=
```

Comment:
`z` and `g` are used several times, it's better to have them in a `variables...` line.